class_allocation.py

```diff
@@ -299,78 +301,66 @@ def ecpc(self):
-                emis_start_w = self.emis_hist.sel(Time=self.start_year_analysis,
-                                                            Region='EARTH')
-                emis_ratio = emis_start_r / emis_start_w
-                path_scaled_0 = emis_ratio * globalpath
-                budget_without_assumptions = path_scaled_0.sum(dim='Time')
-                budget_surplus = budget_left - budget_without_assumptions
-
-                def ecpc_factor(initial_path, f):
-                    '''
-                    Calculates a modified emissions path. Takes scaling factor f as input and
-                    returns a new emissions path by scaling the compensation form (xr_comp) with f
-                    and adding it to the initial emissions path (path_scaled_0).
-                    '''
-                    return initial_path + xr_comp * f
-
-                ecpc = ecpc_factor(path_scaled_0, budget_surplus)
-                ecpc_expanded = ecpc.expand_dims(Discount_factor=[discount],
-                                                 Historical_startyear=[startyear]).to_dataset(name='ECPC')
-                xrs.append(ecpc_expanded)
-
-        xr_ecpc = xr.merge(xrs)
-        self.xr_total = self.xr_total.assign(ECPC = xr_ecpc.ECPC)
+                hist_emissions_rt = (hist_emissions * xr_discount)
+                hist_emissions_wt = (hist_emissions * xr_discount).sel(Region='EARTH')
+                Var_H = (share_popt_past*hist_emissions_wt - hist_emissions_rt).sel(Time=np.arange(startyear, 2020+1)).sum(dim='Time').sel(Region=self.focus_region)          # Historical leftover (or debt if negative)
+
+                for conv_year in self.dim_convyears:
+
+                    # Set up starting variables that will change over iteration
+                    Var_E = (global_emissions_future.sel(Time=2021)*GF_frac)
+                    globe_new = global_emissions_future.sel(Time=2021)
+                    Var_As = (global_emissions_future*population_data.sel(Region=self.focus_region) / population_data.sel(Region='EARTH'))     # Additional allocation because of per capita in that year
+                    Var_L = Var_H - Var_E + Var_As.sel(Time=[2021]).sum(dim='Time')                                                                                 # Current leftover / debt
+                    es = [Var_E]
+
+                    max_time_steps = conv_year-2021
+
+                    for t in range(2100-self.start_year_analysis):
+                        globe_new = global_emissions_future.sel(Time=2022+t)
+                        pop_frac = share_popt.sel(Time=2022+t, Region=self.focus_region)
+                        if t < max_time_steps-1:
+                            Delta_L = Var_L / (max_time_steps - t)
+
+                            # Update variables
+                            Var_E = Delta_L*np.sin((t+1) / max_time_steps*np.pi)*3 + globe_new*(GF_frac*(1-(t+1)/(max_time_steps)) + pop_frac*((t+1)/(max_time_steps)))
```

Comment:
This is quite a long and complex line that can be further broken down for more transparency.